Ccf tag092

prakashngit opened the pull request:
Moving CCF to tag version 092. Further, expanding CCF documentation to include stand alone instructions for build and deploy.

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+# Install CCF Dependencies
+
+CCF/PDO combo has been tested under a scenario where CCF is deployed in a standalone VM. Further, the CCF/PDO combo
+has been tested only for the CCF virtual enclave mode. The dependencies needed to deploy CCF in an Ubuntu 18.04  VM
```
Is it only not tested in HW mode (but should in principle work) or is there any fundamental limitation currently from making it work in HW mode?

prakashngit replied:
Thank you for the question. To make CCF work with HW mode, two changes need to be made 1. sign the CCF enclave applicaiton (a few lines to CMakelists.txt) 2. specify OE configuration. We will add these shortly. (I have tested them in Azure, they work)

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+Copy the transaction processor source files to the CCF src/apps/ folder
+
+mkdir $PDO_SRC/ccf_transaction_processor/CCF/src/apps/pdo_tp
+cp $PDO_SRC/ccf_transaction_processor/transaction_processor/*.* $PDO_SRC/ccf_transaction_processor/CCF/src/apps/pdo_tp/
```
instead of copying, just adding a symlink might be easier?  In fact, couldn't you just add modify the paths in `add_ccf_app` to point to `../transaction_processor`?

Haven't tested either of the simplifications yet as i still struggle with ansible ...

prakashngit replied:
yes, thank you, fixed this!

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+        C. Set the build flags
+
+                $PDO_SRC/ccf_transaction_processor/CCF/build
+                cmake -GNinja -DCOMPILE_TARGETS=virtual -DBUILD_END_TO_END_TESTS=OFF -DBUILD_SMALLBANK=OFF
+                              -DBUILD_TESTS=OFF-DBUILD_UNIT_TESTS=OFF ..
```
```suggestion
                cmake -GNinja -DCOMPILE_TARGETS=virtual -DBUILD_END_TO_END_TESTS=OFF -DBUILD_SMALLBANK=OFF \
                              -DBUILD_TESTS=OFF-DBUILD_UNIT_TESTS=OFF ..
```
 so copy/paste will work

prakashngit replied:
thank you. fixed this.

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+before executing the above comamnd:
+
+        A. In the setup_nosgx.sh file, modify the command `sudo add-apt-repository ppa:ansible/ansible -y` by adding the `-E` option.
+           The modified command looks like `sudo -E add-apt-repository ppa:ansible/ansible -y`. It may also be necesssary to add a `sudo`
```
> But eventually i run into a problem with Python 3.7 ppa setup which fails prey to bug [ansible/ansible#42534](https://github.com/ansible/ansible/issues/42534). (Note the prefix with sudo, even with -E, does not help!)

Almost certainly this is the problem: commenting out the Python 3.7 task makes ansible work as does pre-adding the corresponding apt repo via `sudo add-apt-repository ppa:deadsnakes/ppa` ..

prakashngit replied:
Yes, you are right that `sudo add-apt-repository ppa:deadsnakes/ppa` seems to be necessary while working behind proxies. Both my previous attempts were on machines that already had this repo already added to the source list. Thanks for bringing this up, I will add this to the instructions.

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+
+The following commands starts a single node CCF under the virtual enclave mode.  PDO-TP is the hosted application.
+
+cd $PDO_SRC/ccf_transaction_processor/CCF/build
+python3.7 -m venv env
+source env/bin/activate
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
```
You want to mark up these commands as proper commands; right now they render as a (completely unparseable) single text paragraph

prakashngit replied:
sure, thank you, will fix this.

prakashngit commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+has been tested only for the CCF virtual enclave mode. The dependencies needed to deploy CCF in an Ubuntu 18.04  VM
+with  virtual enclave mode can be installed by running the following command:
+
+$PDO_SRC/ccf_transaction_processor/CCF/getting_started/setup_vm/setup_nosgx.sh.
```
yes, thank you, fixed this!

g2flyer replied:
you only fixed the second part: `setup_nosgx.sh` does not exist anymore; it is called `setup_nodriver.sh` ....

prakashngit replied:
thanks, done!

prakashngit commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+           The modified command looks like `sudo -E add-apt-repository ppa:ansible/ansible -y`. It may also be necesssary to add a `sudo`
+           prefix to the `ansible-playbook ccf-dependencies-no-driver.yml` command appearing in the same file.
+
+        B. In ccf-dependencies-no-driver.yml and ccf-dependencies.yml files (present under the same folder), add the environment option for proxies
```
you are right, I have removed this.

prakashngit commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
+
+        C. Set the build flags
+
+                $PDO_SRC/ccf_transaction_processor/CCF/build
```
Thank you, fixed this.

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+```
+
+The above script works only when the VM is not behind any proxies. If there are proxies, make the following changes
+before executing the above comamnd:
````
```suggestion
before executing the above command:
```

prakashngit replied:
thank you! done

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+
+	C. Add the repo used by ansible scripts for installing python 3.7
+
+		```bash
+		sudo add-apt-repository ppa:deadsnakes/ppa
+		```
````
Something is still wrong in the markup, at least as far as seen in https://github.com/hyperledger-labs/private-data-objects/blob/46fc778cbe3225b2b727315fb060b602667f4e8a/ccf_transaction_processor/Readme.md:  the whole list already seems to be in some quote/verbatim mode and then is is shown literally including the back-quotes ...

prakashngit replied:
thank you, removed all indents.

prakashngit replied:
apparently, it is  possible to preview .md files in vscode, never knew this until now!

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+
+```bash
+cd $PDO_SRC/ccf_transaction_processor/test
+./test.sh <ip-address>
````
This creates temporary files in that directory; would be good to add them to `.gitignore`

prakashngit replied:
thank you, sure

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+
+```bash
+cd $PDO_SRC/ccf_transaction_processor/generate_signing_keys
+./gen_keys.sh <ip-address>
````
This creates new ephemeral files in the repo, would be good to add them to `.gitignore`

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+
+# Share CCF Keys
+
+CCF uses mutually authenticated TLS channels for transactions. Keys are located at
+$PDO_SRC/ccf_transaction_processor/CCF/build/workspace/pdo_tp_common. The network certificate is `networkcert.pem`.
+User public certificate is `user0_cert.pem` and private key is `user0_privk.pem`.
+
+In our usage, CCF users are PDO clients, and PDO client authentication is implemented within the transaction processor itself. Thus, we do not utilize the client authentication feature provided by CCF. However to satisfy the CCF's requirement that only authorized CCF users can submit transactions to CCF, share `user0_cert.pem` and `user0_privk.pem` with all the PDO clients. These two keys must be ...
```
I would have assumed we also would have to somewhere have a trusted version of the CCFs PK (`networkcert.pem`) to authenticate that we are really talking to the right CCF?  Certainly would seem to be crucial for what you say later about signed Payloads of read transaction below?

g2flyer replied:
Actually, you _do_ require that networkcert.pem is also copied to PDO_LEDGER_KEY_ROOT!

g2flyer replied:
btw: seems this copying has to be done _each time_ i restart, even if i already have a workspace?  Is there a way to start the TP on existing keys?

prakashngit replied:
Yes, thank you, you will have to store the network keys as well. And, I would not imagine that there is a way, since the keys are unique to an instance of the CCF enclave.

g2flyer replied:
well, i was wondering about a multiple instances of the nodes but for a single instance of a workspace.  That said, it is true that the whole idea with Raft+SGX really relies on having at any point in time some running enclaves and having all (bootstrap/root) keys/state only memory and never stored/sealed.  Insofar i guess my question didn't really make sense ...

g2flyer commented on ccf_transaction_processor/test/ping_test.py:
```diff
@@ -0,0 +1,55 @@
+
+    num_pings = args.num_pings
+    host = args.host
+    port = 6006
```
given that i have to specify the port for creating the TP, it would have seen more natural that the argument to --host is also of the form <ip:port>. In fact, calling it `--node` in that case would be even easier ....

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+./get_verifying_keys.sh <ip-address>
+```
+
+The read-paylaod-signature feature may be used by PDO clients to establish offline verifiable proof of transaction commits as desired by the PDO smart contract application. Note that for trust purposes, it is recommended that any entity that uses the verifying_key gets it directly from the CCF service using the `get_ledger_verifying_key` rpc.
````
What seems to miss in this doc is then also how to connect the CCF PDO to the PDO "front-end" (EHS, PS).  It is of course somewhat hinted in other documents (e.g., https://github.com/hyperledger-labs/private-data-objects/blob/master/docs/environment.md), but still would be nice to have a section how you could pull it together and, say, run a mock-contract and/or the standard tests?

prakashngit replied:
yes, agreed, I am adding a new section to the document on "Usage". Thank you!

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -6,37 +6,123 @@ https://creativecommons.org/licenses/by/4.0/
-rpc after deployment to generate these keys. Further, the admin must ensure that this key pair is globally committed
-before making the ledger available for PDO usage. This can checked by issuing the rpc again & ensuring that a verifying key gets returned. See https://microsoft.github.io/CCF/users/issue_commands.html for instructions on
-how an rpc can be issued via command line.
+The software is located under $PDO_SRC/transaction_processor/. The TP software is written and tested for CCF tag 0.9.2. Compatability with other CCF versions is not guaranteed.
```
actually, we mention in some docs `$PDO_SOURCE_ROOT`, so this would be more consistent to use here?

prakashngit replied:
thanks, done!

g2flyer replied:
you haven't pushed the corresponding commits, though?

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
+
+A single node CCF instance hosting the PDO TP will be available for business use @ ip-address:6006. Here `ip-address` is the address of this VM that can be used to ping this machine from other VMs (this is our use case). Set ip-address to 127.0.0.1 for local testing. If you are behind a proxy, add ip-address to the list of no_proxy bash env variable.
````
btw: isn't this what you also will have to define (presumably with http:// prefix?) as `PDO_LEDGER_URL`?  If so, worth mentioning ...

g2flyer replied:
Actuallly, might be easiest to define early some env-vars along the lines of 
```
     export PDO_CCF_HOST=172.18.0.1
     export PDO_CCF_PORT=6006
```
and then you can define 
```
export PDO_LEDGER_URL=http://${PDO_CCF_HOST}:${PDO_CCF_PORT}
```
and use `${PDO_CCF_HOST}` and `${PDO_CCF_PORT}` for where you say <ip-address> and 6006?

prakashngit replied:
To your first question, yes, sure i will add a note that URL PDO_LEDGER_URL is related (and say how). I would not want the two env variable option, since PDO clients do not see this issue. I would still prefer the admin to do a little more work, and make it easy for the clients.

g2flyer replied:
well, these environment variables should make it _easier_ than right now as you have a single place where port and ip is defined rather than several?  That certainly what i did as the ip when using docker was not always the same and it was way easier to redefine the env var and have all commands parameterized.  (E.g., i also added a PDO_CCF_BUILD_DIR var which made commands shorter and also enabled me more easily to have the build dir outside of the git repo (primarily for docker reason as i didn't wanted docker images which drag in 3GB of CCF build artifacts [surprising how big the CCF build dir is ....]

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+$PDO_SOURCE_ROOT/ccf_transaction_processor/CCF/build/workspace/pdo_tp_common. The network certificate is `networkcert.pem`. User public certificate is `user0_cert.pem` and private key is `user0_privk.pem`.
+
+In our usage, CCF users are PDO clients, and PDO client authentication is implemented within the transaction processor itself. Thus, we do not utilize the client authentication feature provided by CCF. However to satisfy the CCF's requirement that only authorized CCF users can submit transactions to CCF, share `user0_cert.pem` and `user0_privk.pem` with all the PDO clients. These two keys and the ...
+
```
Might be worth mentioning that you have to do that each time you restart the CCF PDO_TP!  For easy of use for reader and as above paragraph is easy to overlook, you might also want to explicitly add commands, e.g.,
```
     /bin/cp -f ${PDO_CCF_BUILD_DIR}/workspace/pdo_tp_common/user0_cert.pem\
        $PDO_LEDGER_KEY_ROOT/userccf_cert.pem
     /bin/cp -f ${PDO_CCF_BUILD_DIR}/workspace/pdo_tp_common/user0_privk.pem\
        $PDO_LEDGER_KEY_ROOT/userccf_privk.pem
     /bin/cp -f ${PDO_CCF_BUILD_DIR}/workspace/pdo_tp_common/networkcert.pem\
        $PDO_LEDGER_KEY_ROOT/
```?

g2flyer commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+## Generate Signing Keys used to sign Payloads of Read Transactions
+
+Responses to read-transactions include a payload signature, where the signature is generated within PDO-TP.
+The required signing keys must be generated before PDO-TP can be opened up for business from PDO clients.
```
```suggestion
The required signing keys must be generated each time after you start the CCF PDO-TP before it can be opened up for business from PDO clients.
```
Along the lines of the earlier comments on the other keys

g2flyer commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+
+2. Set env PDO_LEDGER_KEY_ROOT and save CCF's network certificate `networkcert.pem` and user keys @ PDO_LEDGER_KEY_ROOT. Note that the user cert and private keys must be named as `userccf_cert.pem` and `userccf_privk.pem` respectively.
+
+3. Do a clean build of PDO
+
+```bash
+cd $PDO_SOURCE_ROOT/build/
+make clean
+make
+```
+A clean build is a an easy way to ensure updated creation of config files and PDO keys that are comptabible with CCF.
+
````
forcing a clean complete build seems excessive.  It is sufficient to do
```
source ${PDO_SOURCE_ROOT}/build/common-config.sh
make -C ${PDO_SOURCE_ROOT}//build force-conf keys
```
for the redefinition of the exports you have in this section, the `source` is probably also not necessary.  If you run it in hardware mode you will though also have to add the target `register` to the `make` command

masomel commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+## Share CCF Keys
+
+CCF uses mutually authenticated TLS channels for transactions. Keys are located at
+$PDO_SOURCE_ROOT/ccf_transaction_processor/CCF/build/workspace/pdo_tp_common. The network certificate is `networkcert.pem`. User public certificate is `user0_cert.pem` and private key is `user0_privk.pem`.
```
I assume these `user0` keys and certs created as part of the CCF build process? It might be worth adding a note about this here so users know where the keys come from.

prakashngit replied:
the keys are created when you start CCF. the keys are unique to the instance of CCF. Thank you.

masomel commented on ccf_transaction_processor/generate_signing_keys/gen_keys.sh:
```diff
@@ -0,0 +1,34 @@
+# Usage ./gen_keys.sh <ip-address-of-CCF>
+
+# Copy KEYs for test
+cp ../CCF/build/workspace/pdo_tp_common/user1_privk.pem .
```
The documentation mentions `user0` keys. Are these `user1` keys different? If so, are they also created at build time? And if so, it might be worth mentioning in the documentation that there are multiple TLS channel key pairs generated at build time.

prakashngit replied:
two additional pairs of user keys get generated as part of deployment  (this is if i use their default start script). I am sure there is a more convoluted way to use their scripts to only allocate one user. but given that we really do not care about publishing user keys for PDO, probably not worth the time. Sure, will add a note that there are two additional sets of user keys. Thank you!

masomel commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+
+A single node CCF instance hosting the PDO TP will be available for business use @ ip-address:6006. Here `ip-address` is the address of this VM that can be used to ping this machine from other VMs (this is our use case). Set ip-address to 127.0.0.1 for local testing. If you are behind a proxy, add ip-address to the list of no_proxy bash env variable.
+
+## Share CCF Keys
```
At first, I thought this section was going to talk about the ledger verifying key. Maybe we should clarify that this section talks about the TLS keys:

```suggestion
## Share CCF TLS Keys
```

prakashngit replied:
thank you, added

masomel commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+
+The CCF port is assumed to be 6006.
+
+## Generate Signing Keys used to sign Payloads of Read Transactions
```
In contrast to the above TLS key section:
```suggestion
## Generate Ledger Signing Keys used to sign Payloads of Read Transactions
```

prakashngit replied:
thank you, added

prakashngit commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+
+## Build CCF & PDO-TP
+
+CCF uses a combination of cmake& ninja to build the applcaition. Execute the following steps to complete the build process:
```
thank you (finally decided to install a spell checker for vscode, hopefully no more!)

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+./get_verifying_keys.sh <ip-address>
+```
+
+The read-paylaod-signature feature may be used by PDO clients to establish offline verifiable proof of transaction commits as desired by the PDO smart contract application. Note that for trust purposes, it is recommended that any entity that uses the verifying_key gets it directly from the CCF service using the `get_ledger_verifying_key` rpc.
````
thank you, fixed

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+
+```bash
+add_ccf_app(
+pdoenc SRCS ../transaction_processor/pdo_tp.cpp
````
pdoenc just is the partial name of the library to be created, the full name will be libpdoenc.virtual.so. Thank you

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,172 @@ Licensed under Creative Commons Attribution 4.0 International License
+ninja
+```
+
+If the VM has a small memory (< 4GB), it might be required to reduce the parallelism in the build process using the `-j` flag for the `ninja` command, say ninja -j2 or even ninja -j1 (the default seems to be ninja -j4).
````
all dependenices, including ninja will be installed you execute the script mentioned in Install CCF Dependencies section.Thank you.

masomel replied:
Got it, thanks!

prakashngit commented on ccf_transaction_processor/transaction_processor/pdo_tp.cpp:
```diff
@@ -49,10 +49,48 @@ namespace ccfapp
 
+        //======================================================================================================
+        // gen_signing_key implementation
+        auto gen_signing_key = [this](Store::Tx& tx, const nlohmann::json& params) {
```
Thank you for the question, unfortunately, AFAIK, there is no way for rpc to issue a local comit and wait until global commit to return the rpc. only way is to check status again to see if commit is successful or not.

masomel replied:
I suppose we could implement the wait loop ourselves, but it seems like that just amounts to doing `get_ledger_verifying_key()` calls anyway... Might be something to consider doing in the future, but since the responses clearly indicate how to poll the ledger by checking the status, I'm fine leaving this as-is for now.

prakashngit commented on ccf_transaction_processor/transaction_processor/pdo_tp.cpp:
```diff
@@ -65,29 +103,19 @@ namespace ccfapp
                 catch(...){
                     return make_error(
-                        jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Unable to locate ledger Key");
+                        HTTP_STATUS_BAD_REQUEST, "Unable to locate ledger Key despire global commit");
```
Thank you, fixed the typo.
The error code HTTP_STATUS_BAD_REQUEST seems to be CCF recommended way to return error (see https://microsoft.github.io/CCF/master/developers/logging_cpp.html).  The code does not seem to have much meaning, I guess we need to interpret only with the message, given that all errors will have the same code. That said, it is definitely possible to add a new code by adding to http_parser.h (http-parser is a third party component used by CCF), but at this point, I suggest that we follow the style adopted by CCF examples.

masomel replied:
I see. Changing a third-party CCF component seems a bit excessive for us to get more meaningful error codes. In most cases the `HTTP_STATUS_BAD_REQUEST` actually works.

prakashngit commented on ccf_transaction_processor/test/test.sh:
```diff
@@ -0,0 +1,35 @@
+# limitations under the License.
+
+
+# Usage ./test.sh <ip-address-of-CCF>
```
may be, i can come back if we add more tests.

masomel replied:
I think it would be nice to be consistent with the other PDO script naming conventions we have, but I won't push for it too much right now.

cmickeyb commented:
@prakashngit once you have addressed the feedback you have received so far, please squash the commits and force push the result.

prakashngit commented:
> @prakashngit once you have addressed the feedback you have received so far, please squash the commits and force push the result.

Thank you. Done.

bvavala commented on ccf_transaction_processor/generate_signing_keys/gen_keys.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Usage ./gen_keys.sh <ip-address-of-CCF>
+
+# Copy KEYs for test
```
This seems a generic script, yet some operations are for "tests".
If additional work is required to separate common operations from tests, it's better to add a detailed TODO.

prakashngit replied:
thank you, some of the comments in this script are wrong (blame my copy paste). I am fixing them. this is not a test script. there is no plan to add more tests, other than testing ccf via using with along with pdo (via run tests). Thanks.

bvavala commented on ccf_transaction_processor/generate_signing_keys/get_verifying_keys.sh:
```diff
@@ -0,0 +1,34 @@
+# Usage ./get_verifying_keys.sh <ip-address-of-CCF>
+
+# Copy KEYs for test
+cp ../CCF/build/workspace/pdo_tp_common/user1_privk.pem .
```
(same here, tests vs generic ops)

prakashngit replied:
thank you, same answer as above.

bvavala commented on ccf_transaction_processor/generate_signing_keys/key_gen.py:
```diff
@@ -0,0 +1,46 @@
+    
+    host = args.host
+    port = 6006
+    cert = "./user1_cert.pem"
```
here user1 is hardcoded, and there is no mention of tests. Will this require any refactoring later?

prakashngit replied:
thank you, I am changing it to user0. No refactoring later. CCF users are named user0, user1 etc. As of now the deployemnt script by default generates 3 CCF users user0/1/2. given that in the documentation, we ask pdo clients to take the role of user0, i am changing these scripts also to use user0 (as noted in an earlier response, we might be able to deploy ccf with exactly one user, but given that in pdo we manage user identities within the TP itself, I am not going to see how to fix this now). Thanks !

bvavala commented on ccf_transaction_processor/generate_signing_keys/key_gen.py:
```diff
@@ -0,0 +1,46 @@
+    client = CCFClient(host, port, cert=cert, key=key, ca = cafile, format=format, prefix="users", description="none", \
+        version="2.0",connection_timeout=3, request_timeout=3)
+
+    r=client.rpc("generate_signing_key_for_read_payloads", dict())
```
looking at the CCFClient definition, I think you're missing some error handling here, specifically a try/catch for exceptions.

prakashngit replied:
Thank you, so the line below this checks if the response has an error or not. basically, if the error comes from CCF iteslf, the code will report it. Yes, we do not catch an error while trying to submit the rpc in this script, though when we submit rpcs in PDO clients, I do check that as well. This script is basically a one time manual script, so I dont think it makes much difference. Thanks!

bvavala commented on ccf_transaction_processor/generate_signing_keys/key_read.py:
```diff
@@ -0,0 +1,46 @@
+
+    host = args.host
+    port = 6006
+    cert = "./user1_cert.pem"
```
(same as above, hardcoded users vs generic util)

prakashngit replied:
Thank you, please see above.

cmickeyb commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
+
````
i ran the first time without a node parameter and i had a three node system setup. when i went back and specified an ip address (e.g. for a different interface), the script crashed.

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
+
````
thank you, i will add a new start script to wrap these.

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+source env/bin/activate
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
````
sure, i will default it to 127.0.0.1, and give an option to pass anything else as an argument in the wrapper script

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+pdoenc SRCS ../transaction_processor/pdo_tp.cpp
+../transaction_processor/verify_signatures.cpp
+)
+```
````
Thank you, yes, I was going to include our own CMakeLists.txt as part of the docker PR, instead of having to edit this.

prakashngit commented on ccf_transaction_processor/Readme.md:
```diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+If the VM has a small memory (< 4GB), it might be required to reduce the parallelism in the build process using the `-j` flag for the `ninja` command, say ninja -j2 or even ninja -j1 (the default seems to be ninja -j4).
+
+If build is successful, the libpdoenc.virtual.so library is created inside the build folder (ignore any other targets that get created)
+
```
thank you, yes definitely, most of the install commands can be compressed into one shell script, but my understanding after reading pdo build script was that we would like to detail all the steps and provide explanation, I am happy to change if that makes sense here.

prakashngit commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+cd $PDO_SOURCE_ROOT/ccf_transaction_processor/CCF/build
+cmake -GNinja -DCOMPILE_TARGETS=virtual -DBUILD_END_TO_END_TESTS=OFF -DBUILD_SMALLBANK=OFF \
+                -DBUILD_TESTS=OFF-DBUILD_UNIT_TESTS=OFF ..
+```
````
thank you, sure, that is cleaner, will do.

cmickeyb commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+source env/bin/activate
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
````
right now the network uses my personal ssh key at some point. this behavior is not appropriate. if an ssh key is needed for the client it should be created and we should ensure that it is used.

cmickeyb commented on ccf_transaction_processor/Readme.md:
````diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+
+```bash
+cd $PDO_SOURCE_ROOT/ccf_transaction_processor/test
+./test.sh <ip-address>
````
when i tried this i got a python exception, about missing module infra. i have activated the virtual environment.

prakashngit replied:
may i ask if you did `cd`  into the directory before executing the test.sh. Infra is part of ccf's test module. There is hardcoding of paths, I can't think of any other reason why infra is not found. venv activation is taken care of by the test.sh script as well.

cmickeyb commented on .gitignore:
```diff
@@ -4,3 +4,7 @@ docker/sgx
+ccf_transaction_processor/test/infra
+ccf_transaction_processor/test/*.pem
+ccf_transaction_processor/generate_signing_keys/infra
+ccf_transaction_processor/generate_signing_keys/*.pem
```
do we need these changes? some of these directories no longer exist.

prakashngit replied:
Not needed after the refactoring. Will remove them. Thank you.

cmickeyb commented on ccf_transaction_processor/etc/sample_key.pem:
```diff
@@ -0,0 +1,39 @@
+B+uxRXxT2DmbG4VCuydGCWT0JBX6yfpHHPsulgbUZRXBq4ftkEst9SHT7JbjGArP
+ZWa2/FyApo73LM3pE9izmFOlMHgnnKGZw7KOOm5GhjH2ZF+6b7kZVOL1/FTDxJqd
+dMk3oj+hUnuOO/OVZgjuL7HCf1YlIXyJLZt42r2xeVYS7otYU0A=
+-----END RSA PRIVATE KEY-----
```
is this file still needed? i don't believe we use it anywhere.

prakashngit replied:
Thank you for the question. The key is used to sign the ccf enclave app (pdo_tp), which is necessary we want to run ccf inside actual enclaves rather than the virtual mode. (the current ccf enclave app must run in actual HW mode as well, we just need to test this, so we do not claim in this PR, but software-wise, everything seems complete)

cmickeyb replied:
thanks. might want to rename this : sample_enclave_signing_key.pem

cmickeyb commented on ccf_transaction_processor/README.md:
```diff
@@ -0,0 +1,208 @@
+
+A. In the setup_nosgx.sh file, modify the command
+`sudo add-apt-repository ppa:ansible/ansible -y`
+by adding the `-E` option. The modified command looks like
```
shouldn't all of these be apt-add-repository rather than add-apt-repository?

prakashngit replied:
Hi Mic, Thank you for the question. No, it is add-apt-repository (http://manpages.ubuntu.com/manpages/trusty/man1/add-apt-repository.1.html). But I see another issue to fix, the file name is setup_nodriver and not setup_nosgx (that was for some older version of ccf, I thought i had fixed this, not sure)

g2flyer commented on ccf_transaction_processor/scripts/start_cchost.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
```
actually, why don't you take the same logic as used in common_service.sh for eservice. that also captures if the service doesn't start (which you currently do not catch, i.e., you still exit with success even when service immediately dies ...)

g2flyer replied:
btw: the common failure for start seems to be when the network key already exists.  Maybe worth testing for that explicit case?   As long as we have only one host, you could just remove it?

prakashngit replied:
Thank you for the comments. Regarding the last one, the start_cchost is invoked only from within start_ccf_network.sh, not independently. Any prior network certificates will be removed before invoking start_cchost. There is the usual try, die to report invocation errors, but i need to see what happens if the service does not start.

g2flyer replied:
hmm, iff start_cchost is never called standalone, why even make it a script? If it is a separate script, then i think it should test for unexpected failures ...

g2flyer commented on ccf_transaction_processor/scripts/fetch_ledger_authority.py:
```diff
@@ -0,0 +1,120 @@
+
+    fetch_ledger_authority(user_client, options, config)
+
+    LOG.info('successfully generated ledger authority')
```
```suggestion
    LOG.info('successfully fetched ledger authority key')
```

prakashngit replied:
Thank you, fixed this

g2flyer commented on ccf_transaction_processor/scripts/fetch_ledger_authority.py:
```diff
@@ -0,0 +1,120 @@
+
+    parser.add_argument('--ccf-config', help='Name of the CCF configuration file', default = default_config, type=str)
+    parser.add_argument('--user-name', help="Name of the user being added", default = "user0", type=str)
+    parser.add_argument("--output-file", help="Name of the file where the key will be saved", default = default_output, type=str)
```
what is the default here? I would have assumed that not providing an argument would print it to stdout or alike but that does not happen ...

prakashngit replied:
Thank you for the question. I guess you are asking about the default_output, which is the file ledger_authority_pub.pem located under the keys folder. This is defined in line 63 of the same file.

g2flyer commented on ccf_transaction_processor/README.md:
````diff
@@ -0,0 +1,208 @@
+
+```bash
+export PDO_LEDGER_TYPE=ccf
+export PDO_LEDGER_URL=http://ip-address:6006
````
Hmm, by default our scripts/settings start the server only on 127.0.0.1, so the instruction here might be confusing ...

Is there no way that CCF might just listen on all interfaces?

If not, you will have to add some text here what you have to do if it is not localhost (e.g., for my test case where ccf was running on host and ehs/test in docker, i've changed cchost.toml to point to the address on the docker network (172.20.0.1), restarted network  and then used that in the test ...)

prakashngit replied:
Thank you for the question. By default, ccf is started on HOSTNAME:6006, where HOSTNAME is one of the env variables that must be set before configuring CCF. We also recommend setting HOSTNAME to the ipaddress to avoid issues that arise while converting names to ip.

g2flyer replied:
I don't think your new text clarifies much.  Most folks would probably read that as the main (non-localhost) ip address of the host, not necessarily the default 127.0.0.1?

prakashngit replied:
Thank you. I changed this to 
```bash
export PDO_LEDGER_TYPE=ccf
export PDO_LEDGER_URL=http://ccf-ip-address:6006
```

Note that ccf-ip-address is the ip-address that was set in the env variable HOSTNAME (see above) during CCF
deployment. 
========================================
HOSTNAME is one of the env that we expect users to set before deploying CCF (we also say that the user should set it to an ip rather than name. This can be set to `127.0.0.1` as well. The env is already in PDO. Setting it to a name like localhost might also work (since there is some conversation within scripts, but I am not sure if this conversation works in all cases, so we recommend to set it explicitly to the ip)

g2flyer commented on ccf_transaction_processor/README.md:
```diff
@@ -0,0 +1,208 @@
+
+Here, ip-address is the address of the node which hosts the CCF instance.
+
+2. Set env PDO_LEDGER_KEY_ROOT and save CCF's network certificate
```
what does PDO_LEDGER_KEY_ROOT have to be set to?!  This whole paragraph is a bit rough reading. You might also refernce back to the "Share CCF (TLS) Authentication Keys" section for more context?

prakashngit replied:
Thank you for the question. PDO_LEDGER_KEY_ROOT is the path to the folder where the keys must be saved. This paragraph is a brief of what was already explained earlier in the same README.md. Sorry for any lack of clarity, I have clarified better now.

g2flyer commented on ccf_transaction_processor/README.md:
````diff
@@ -0,0 +1,208 @@
+
+```bash
+cd $PDO_SOURCE_ROOT/build/
+make clean
````
as mentioned in previous review, forcing a clean build seems excessive; just to switch between ledger types, following seems to be sufficient
```
source ${PDO_SOURCE_ROOT}/build/common-config.sh
make -C ${PDO_SOURCE_ROOT}//build force-conf keys
```
for the redefinition of the exports you have in this section, the source is probably also not necessary. If you run it in hardware mode you will though also have to add the target register to the make command.

(In fact, in the pdo_build docker image it seems it works even without any step after a prior build (with ledger type and url defined for sawtooth and just redefining the two variables is sufficient.)

prakashngit replied:
Thank you, I am adding these instructions as alternatives to make clean.

cmickeyb commented on build/__tools__/expand-config:
```diff
@@ -19,11 +19,57 @@ Return the number of registered endpoints in the exit status
 import re
 from string import Template
-import pdo.common.config as pconfig
+
```
because we can run expand config without installing all of PDO in the CCF container. Otherwise, you'd have to install all of PDO in order to pick up two functions.

prakashngit commented on ccf_transaction_processor/transaction_processor/pdo_tp.h:
```diff
@@ -113,9 +116,10 @@ namespace ccfapp
     class TransactionProcessor : public ccf::UserRpcFrontend
     {
     private:
-        TPHandler ping_handler;
```
thank you for the comment. sure will do.

prakashngit commented on ccf_transaction_processor/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+#
```
Thank you, added.

prakashngit commented on ccf_transaction_processor/transaction_processor/pdo_tp.h:
```diff
@@ -58,6 +60,7 @@ namespace ccfapp
     //method that read the ledger verifying key. The very first invocation of this rpc will
     //create the keys and globally commit it.
 
+    static constexpr auto GEN_SIGNING_KEY = "generate_signing_key_for_read_payloads";
```
Thank you, fixed

prakashngit commented on ccf_transaction_processor/README.md:
````diff
@@ -0,0 +1,208 @@
+
+```bash
+cd $PDO_SOURCE_ROOT
+git submodule init
+git submodule update
````
Thank you. Fixed!

prakashngit commented on ccf_transaction_processor/Makefile:
```diff
@@ -0,0 +1,131 @@
+
+build-ccf :
+	mkdir -p $(CCFDIR)/build
+	cd $(CCFDIR)/build; cmake -GNinja \
```
Thank you for the note. Good to learn this, will ensure this in future.

g2flyer replied:
hmm, i did fix all of them in my makefile patch i've sent you, but you still have a lot of them left?

prakashngit commented on ccf_transaction_processor/Makefile:
```diff
@@ -0,0 +1,131 @@
+		-DBUILD_TESTS=OFF \
+		-DBUILD_UNIT_TESTS=OFF \
+		..
+	cd $(CCFDIR)/build; ninja; ninja install
```
Thank you, I will add the -j option back.

prakashngit commented on ccf_transaction_processor/Makefile:
```diff
@@ -0,0 +1,131 @@
+
+SCRIPTDIR ?= $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
+CCFDIR := $(SCRIPTDIR)/CCF
+DSTDIR ?= $(PDO_INSTALL_ROOT)
+CCFDSTDIR ?= $(PDO_INSTALL_ROOT)/opt/pdo/ccf
```
Thanks, I thought I had got all of them, you proved me wrong! Fixed it.

cmickeyb commented on ccf_transaction_processor/Makefile:
```diff
@@ -77,8 +83,7 @@ clean-install :
-	@ echo create configuration files for cchost
-	@ . $(abspath $(DSTDIR)/bin/activate) ; \
+	@ . $(abspath $(CCFDSTDIR)/bin/activate) ; \
 		$(CNFGEN) --template $(notdir $<) --template-directory $(dir $<) \
```
We really should use CCFDSTDIR everywhere it is appropriate (see the macros at the top of the file for example). This is not a show stopper especially since i'm sure we'll keep revising this makefile for awhlie.

g2flyer replied:
my patch i've sent to Prakash with which i could make it work actully did that ;-)

prakashngit replied:
applied patch, Thank you!

cmickeyb commented:
start_cchost should be modified to allow rejoining a network.

On Mon, Jul 20, 2020 at 3:23 PM prakashngit <notifications@github.com>
wrote:

> *@prakashngit* commented on this pull request.
> ------------------------------
>
> In ccf_transaction_processor/scripts/start_cchost.sh
> <https://github.com/hyperledger-labs/private-data-objects/pull/282#discussion_r457725153>
> :
>
> > @@ -0,0 +1,47 @@
> +#!/bin/bash
>
> Thank you for the comments. Regarding the last one, the start_cchost is
> invoked only from within start_ccf_network.sh, not independently. Any prior
> network certificates will be removed before invoking start_cchost. There is
> the usual try, die to report invocation errors, but i need to see what
> happens if the service does not start.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperledger-labs/private-data-objects/pull/282#discussion_r457725153>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADMIUGXRSTO5UDPDAHC3GTR4S7URANCNFSM4M6EXGNQ>
> .
>

g2flyer commented on ccf_transaction_processor/Makefile:
```diff
@@ -25,6 +25,12 @@ ifndef SGX_MODE
 $(error Incomplete configuration, SGX_MODE is not defined)
 endif
 
+ifdef NINJA_OPTION
+NINJA_J_OPTION = $(NINJA_OPTION)
+else
+NINJA_J_OPTION = 4
+endif
```
You could have written that way more compact as `NINJA_J_OPTION ?= 4`.

More confusing though, why do you expect `NINJA_OPTION` and then define `NINJA_J_OPTION`? If you really want to allow only override of -j, then you should also make the external variable NINJA_J_OPTION.  Personally, i would have preferred the more general NINJA_OPTIONS i had in the patch i've sent you: more versatile. Also, why are you passing it only to ninja but not, say, ninja install?

g2flyer commented on ccf_transaction_processor/Makefile:
```diff
@@ -0,0 +1,136 @@
+$(error Incomplete configuration, SGX_MODE is not defined)
+endif
+
+ifdef NINJA_OPTION
+NINJA_J_OPTION = $(NINJA_OPTION)
+else
+NINJA_J_OPTION = 4
+endif
```
You could have written that way more compact as NINJA_J_OPTION ?= 4.

More confusing though, why do you expect NINJA_OPTION and then define NINJA_J_OPTION? If you really want to allow only override of -j, then you should also make the external variable NINJA_J_OPTION. Personally, i would have preferred the more general NINJA_OPTIONS i had in the patch i've sent you: more versatile. Also, why are you passing it only to ninja but not, say, ninja install?

prakashngit replied:
yes, thank you, (like we discussed), I missed the patch, I am applying the patch now. Thanks again.

g2flyer commented on ccf_transaction_processor/Makefile:
```diff
@@ -0,0 +1,136 @@
+		-DBUILD_TESTS=OFF \
+		-DBUILD_UNIT_TESTS=OFF \
+		..
+	cd $(CCFDIR)/build; ninja -j $(NINJA_J_OPTION); ninja install
```
Hmm, you still have the `;` instead of the more robust `&&`.  Note that my patched Makefile i've sent you fixed all of them (as it did remove/cleanup the DSTDIRs ..)

cmickeyb replied:
this is a small thing and is not how we do it in the PDO Makefile. the && is far more important if there might be a failure in the left side which for these there is no chance of failure. if you have it in a patch.. fine. but there are more important things to fix.

prakashngit replied:
patch applied, thank you! (missed it originally, my bad)

g2flyer replied:
> there might be a failure in the left side which for these there is no chance of failure.

Not completely true, see the ninja sequence.  In that case, though, what still helps you of course is that if the left (middle) case fails, the right side should presumably fail as well (exception might be if default build calls failing tests which the install probably won't call).

cmickeyb replied:
that was the one and only place where it was useful and in fact it isn't needed there either if we do the ninja install as the only command. @g2flyer would you please make the pdo build makefile consistent & get a PR that changes the ';' to '&&' there as well?

g2flyer replied:
> @g2flyer would you please make the pdo build makefile consistent & get a PR that changes the ';' to '&&' there as well?

Will do.  I'm already working on a PR for sgx 2.10 (need it for FPC and though i also bump PDO, also as there a few left-over inconsistencies), so depending on how big the change and whether i already have to touch makefiles is might piggy-back it (as separate commit) or create a second PR ...

cmickeyb commented:
agreed in this case. though my question for prakash was whether this
sequence is actually necessary. can we just call "ninja -j
$(NINJA_J_OPTION) install" in which case we're back to "it just doesn't
matter"

On Tue, Jul 21, 2020 at 9:28 AM Michael Steiner <notifications@github.com>
wrote:

> *@g2flyer* commented on this pull request.
> ------------------------------
>
> In ccf_transaction_processor/Makefile
> <https://github.com/hyperledger-labs/private-data-objects/pull/282#discussion_r458229509>
> :
>
> > +
> +all: environment keys config install
> +
> +build: build-ccf build-pdo-tp
> +
> +build-ccf :
> +	mkdir -p $(CCFDIR)/build
> +	cd $(CCFDIR)/build; cmake -GNinja \
> +		-DCOMPILE_TARGETS=virtual \
> +		-DCMAKE_INSTALL_PREFIX=${DSTDIR}/opt/pdo/ccf \
> +		-DBUILD_END_TO_END_TESTS=OFF \
> +		-DBUILD_SMALLBANK=OFF \
> +		-DBUILD_TESTS=OFF \
> +		-DBUILD_UNIT_TESTS=OFF \
> +		..
> +	cd $(CCFDIR)/build; ninja -j $(NINJA_J_OPTION); ninja install
>
> there might be a failure in the left side which for these there is no
> chance of failure.
>
> Not completely true, see the ninja sequence. In that case, though, what
> still helps you of course is that if the left (middle) case fails, the
> right side should presumably fail as well (exception might be if default
> build calls failing tests which the install probably won't call).
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyperledger-labs/private-data-objects/pull/282#discussion_r458229509>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADMIUFXQBCBDU4ASAAYFWTR4W6ZLANCNFSM4M6EXGNQ>
> .
>

prakashngit commented:
> agreed in this case. though my question for prakash was whether this sequence is actually necessary. can we just call "ninja -j $(NINJA_J_OPTION) install" in which case we're back to "it just doesn't matter"
> […](#)
> On Tue, Jul 21, 2020 at 9:28 AM Michael Steiner ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In ccf_transaction_processor/Makefile <[#282 (comment)](https://github.com/hyperledger-labs/private-data-objects/pull/282#discussion_r458229509)> : > + +all: environment keys config install + +build: build-ccf build-pdo-tp + +build-ccf : + mkdir -p $(CCFDIR)/build + cd $(CCFDIR)/build; cmake -GNinja \ + -DCOMPILE_TARGETS=virtual \ + -DCMAKE_INSTALL_PREFIX=${DSTDIR}/opt/pdo/ccf \ + -DBUILD_END_TO_END_TESTS=OFF \ + -DBUILD_SMALLBANK=OFF \ + -DBUILD_TESTS=OFF \ + -DBUILD_UNIT_TESTS=OFF \ + .. + cd $(CCFDIR)/build; ninja -j $(NINJA_J_OPTION); ninja install there might be a failure in the left side which for these there is no chance of failure. Not completely true, see the ninja sequence. In that case, though, what still helps you of course is that if the left (middle) case fails, the right side should presumably fail as well (exception might be if default build calls failing tests which the install probably won't call). — You are receiving this because you were assigned. Reply to this email directly, view it on GitHub <[#282 (comment)](https://github.com/hyperledger-labs/private-data-objects/pull/282#discussion_r458229509)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADMIUFXQBCBDU4ASAAYFWTR4W6ZLANCNFSM4M6EXGNQ> .

prakashngit commented:
I have applied Michael's patch to the makefile. The ninja install command seems to do the installation as well when I try. but i cant find any documentation on this. I have left them as two separate commands.

cmickeyb commented:
> I have applied Michael's patch to the makefile. The ninja install command seems to do the installation as well when I try. but i cant find any documentation on this. I have left them as two separate commands.

i don't understand this comment. ninja is basically make and the install forces all dependencies to be built. since we only really care about the install (in fact... we might want to start using their binary distribution instead of building). i agree that its not a big issue either way.

cmickeyb commented:
> Hmm, i still think some of my conerns apply: using && instead of ; in makefiles should be the way to go 
Both ways are close enough to the same thing & I don't think it makes enough difference in the particular case to warrant even the time for this discussion. I agree that with the patch it should just be applied. And I also believe that we should be consistent across all of our makefiles either way (so @g2flyer please update the PDO makefiles for this style).

> that scripts should have more error checking^1, in particular if we launch a daemon process 
Again, completely agree. Feel free to submit a PR with the updates. However, I think the first thing to do is to verify that we can add hosts to a network with start_cchost.sh. Then we can clean it up & make it more robust.

g2flyer commented:
> > that scripts should have more error checking^1, in particular if we launch a daemon process
> 
> Again, completely agree. Feel free to submit a PR with the updates. However, I think the first thing to do is to verify that we can add hosts to a network with start_cchost.sh. Then we can clean it up & make it more robust.

Well, i did include (ok, not as clickable suggestion) the very simple fix of adding `set -e` which would catch some errors right away. checking that the background starts is a bit more involved (although we have the pattern for ehs).  I can add that as a separate pr once this is merged.  Regarding order, i'd argue having some error checking already in place will actually make (the development of) extending it easier, but i'm definitively not in a rush (in fact some other work-items probably should have priority ...  ;-)
